### PR TITLE
Refactor setup.py

### PR DESCRIPTION
**Pull request by ssbarnea**
Removes all workarounds from setup.py in order to avoid bugs like #590

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -15,7 +15,11 @@ deps =
   pep8-naming
   wheel
 commands = nosetests []
-passenv = HOME
+passenv =
+    CURL_CA_BUNDLE
+    HOME
+    REQUESTS_CA_BUNDLE
+    SSL_CERT_FILE
```
Comment (ssbarnea) [inline tox.ini]: Before someone asks: these are needed if SSL proxies, othewise testing would fail with failure to validate certificates.

File: .travis.yml [hunk 2]
```diff
@@ -150,3 +150,9 @@ install:
 
 script:
 - tox -v
+# test that install and uninstall from source works
```
Comment (webknjaz) [inline .travis.yml]: This shouldn't be mangled into all jobs but should be its own job wrapped into a separate toxenv.

File: setup.cfg [hunk 3]
```diff
@@ -20,6 +20,7 @@ project_urls =
     Source Code = https://github.com/ansible/ansible-lint
 description = Checks playbooks for practices and behaviour that could potentially be improved
 long_description = file: README.rst
+long_description_content_type = text/x-rst
```
Comment (webknjaz) [inline setup.cfg]: Not really needed as it's the default but :+1:

File: tox.ini (hunk 1 above)
Comment (webknjaz) [reply tox.ini]: Instead of a comment in PR, the best practice is to put a comment in the file itself.

File: tox.ini [hunk 4]
```diff
@@ -31,10 +35,21 @@ usedevelop = True
...
+    twine>=1.13.0
+    wheel>=0.31.0
+
 usedevelop = False
 # Ref: https://twitter.com/di_codes/status/1044358639081975813
 commands =
-    python -m setup checkdocs check --metadata --restructuredtext --strict --verbose
-    twine check .tox/dist/*
+    rm -rf {toxinidir}/dist
+    python -m pep517.build \
```
Comment (webknjaz) [inline tox.ini]: 1) this should be a separate testenv, you can then reuse parts of it in other places. also, you can specify deps between testenvs.
2) this is achievable by using `isolated_build = true` so not sure why you want to invoke it explicitly.

File: tox.ini [hunk 5]
```diff
@@ -31,10 +35,21 @@ usedevelop = True
...
+    setuptools>=39.0
+    twine>=1.13.0
+    wheel>=0.31.0
+
 usedevelop = False
 # Ref: https://twitter.com/di_codes/status/1044358639081975813
 commands =
-    python -m setup checkdocs check --metadata --restructuredtext --strict --verbose
-    twine check .tox/dist/*
+    rm -rf {toxinidir}/dist
```
Comment (webknjaz) [inline tox.ini]: This is not cross-platform. Better do
```suggestion
    python -c "import path; path.Path('{toxinidir}/dist').rmtree_p()"
```

File: tox.ini [hunk 6]
```diff
@@ -31,10 +35,21 @@ usedevelop = True
 
 [testenv:metadata-validation]
 deps =
-    collective.checkdocs
-    twine
+    collective.checkdocs>=0.2
+    pep517 >= 0.5.0
```
Comment (webknjaz) [inline tox.ini]: plz, keep the same style of spacing around comparison operators.

File: setup.py [hunk 7]
```diff
@@ -1,150 +1,5 @@
-#! /usr/bin/env python
-"""Ansible-lint distribution package setuptools installer."""
-
+#!/usr/bin/env python
```
Comment (webknjaz) [inline setup.py]: Plz, don't remove the space after `#!`, it's there for readability.
```suggestion
#! /usr/bin/env python
```

File: setup.py [hunk 8]
```diff
@@ -1,150 +1,5 @@
...
-    import setuptools_scm.version  # only present during setup time
-    IS_PYPI_UPLOAD = os.getenv('PYPI_UPLOAD') == 'true'
-    return (
-        '' if IS_PYPI_UPLOAD
-        else setuptools_scm.version.get_local_node_and_date(version)
-    )
-
-
-setup_params['use_scm_version'] = {
-    'local_scheme': cut_local_version_on_upload,
```
Comment (webknjaz) [inline setup.py]: Removing this breaks uploads to Test PyPI.

File: tox.ini (hunk 5 above)
Comment (ssbarnea) [reply tox.ini]: > This is not cross-platform. Better do

While 3rd party module are you referring to because this is clearly not part of standard library. I am not sure if for the sake of portability we need to introduce a new dependency.

Can you please elaborate? What is the library? Is it available for py27...py37?

File: .travis.yml (hunk 2 above)
Comment (ssbarnea) [reply .travis.yml]: We need to test this with all supported python versions but I also want to avoid exploding the number of environments we have (extra footprint). For the moment this should do the trick, we can improve it later. Lets not forget we are trying to fix an almost critical bug #590, not rewrite the entire product. I will be more than happy to add extra improvement latest if you can review them.

File: tox.ini (hunk 5 above)
Comment (webknjaz) [reply tox.ini]: `path.py` and yes, it works on 2.7

File: setup.cfg [hunk 9]
```diff
@@ -64,7 +65,7 @@ keywords =
 use_scm_version = True
 python_requires = >=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*
 package_dir =
-    = lib
+    =lib
```
Comment (webknjaz) [inline setup.cfg]: this is not really needed

File: setup.py [hunk 10]
```diff
@@ -1,150 +1,23 @@
...
 
-
+setup_params = {}
 setup_params['use_scm_version'] = {
     'local_scheme': cut_local_version_on_upload,
 }
 
 
-__name__ == '__main__' and setuptools.setup(**setup_params)
+__name__ == "__main__" and setuptools.setup(**setup_params)
```
Comment (webknjaz) [inline setup.py]: Why did you change the style?

File: .travis.yml (hunk 2 above)
Comment (webknjaz) [reply .travis.yml]: FWIW the quality of this PR is unacceptable, it breaks compatibility with supported envs and mixes up several completely separate changes. So this at least should go into existing testenv in tox where it belongs, not into a selective CI service config.

File: setup.py [hunk 11]
```diff
@@ -1,147 +1,20 @@
...
 def cut_local_version_on_upload(version):
+    """Generate a PEP440 local version if uploading to PyPI."""
     import os
     import setuptools_scm.version  # only present during setup time
-    IS_PYPI_UPLOAD = os.getenv('PYPI_UPLOAD') == 'true'
+
+    IS_PYPI_UPLOAD = os.getenv('PYPI_UPLOAD') == 'true'  # set in tox.ini
     return (
-        '' if IS_PYPI_UPLOAD
+        ""
```
Comment (webknjaz) [inline setup.py]: Plz, don't restructure this, minimize the diff.

File: setup.py [hunk 12]
```diff
@@ -1,147 +1,20 @@
...
+        if IS_PYPI_UPLOAD
         else setuptools_scm.version.get_local_node_and_date(version)
     )
 
-
+setup_params = {}
```
Comment (webknjaz) [inline setup.py]:
```suggestion

setup_params = {}
```

File: tox.ini [hunk 13]
```diff
@@ -35,10 +37,18 @@ usedevelop = True
 
 [testenv:metadata-validation]
 deps =
-    collective.checkdocs
-    twine
+  collective.checkdocs >= 0.2
```
Comment (webknjaz) [inline tox.ini]: Plz don't change the indentation.

File: setup.py (hunk 10 above)
Comment (ssbarnea) [reply setup.py]: Leftovers, fixing them. You can blame "black PTSD" for using double quotes. BTW, You should consider activating black on ansible-lint so we do not have to waste time discussing about formatting.

File: tox.ini (hunk 13 above)
Comment (ssbarnea) [reply tox.ini]: I am afraid you cannot blame me for *FIXING* the identation. That makes the indentation on tox.ini deps consistent with previous deps line. Looks at the file, its a mix of 2 and 4 identations.

File: setup.py (hunk 10 above)
Comment (webknjaz) [reply setup.py]: No, black is still bad.

File: tox.ini (hunk 13 above)
Comment (webknjaz) [reply tox.ini]: it's mostly 4 spaces. this PR is not about reformatting things. so yes, you're breaking it now.

**Comment by webknjaz** [comment]
We cannot accept this atm but this example is good for reference for what we can do in the future. Closing, for now, we can reopen it later once somebody has time to work on fixing the shim or its removal is approved by a higher power.

File: setup.py [hunk 14]
```diff
@@ -1,147 +1,19 @@
...
-    str(k): str(v)
-    for k, v in declarative_setup_params['options']['package_dir'].items()
-}
-
-setup_params = dict(setup_params, **declarative_setup_params['metadata'])
-setup_params = dict(setup_params, **declarative_setup_params['options'])
-
-
 def cut_local_version_on_upload(version):
+    """Generate a PEP440 local version if uploading to PyPI."""
```
Comment (webknjaz) [inline setup.py]: This is not exactly true. It's more like "cut off a local version on upload"

File: setup.cfg [hunk 15]
```diff
@@ -20,6 +20,7 @@ project_urls =
   Source Code = https://github.com/ansible/ansible-lint
 description = Checks playbooks for practices and behaviour that could potentially be improved
 long_description = file: README.rst
+long_description_content_type = text/x-rst
```
Comment (webknjaz) [inline setup.cfg]: FTR this part of the PR will be accepted if submitted separately.
